{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/michael.mitter/.conda/envs/ml/lib/python3.7/site-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as tm\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sbn\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = pd.read_csv(\"../../data/hg19_genes/gencode.v38lift37.basic.annotation.gtf.gz\", comment=\"#\", sep=\"\\t\", header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>5</th>\n",
       "      <th>strand</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_1\"; gene_type \"tran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_1\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_1\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_1\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_1\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878198</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878199</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878200</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878201</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878202</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1878203 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom   source        type  start    end  5 strand  7  \\\n",
       "0        chr1   HAVANA        gene  11869  14409  .      +  .   \n",
       "1        chr1   HAVANA  transcript  11869  14409  .      +  .   \n",
       "2        chr1   HAVANA        exon  11869  12227  .      +  .   \n",
       "3        chr1   HAVANA        exon  12613  12721  .      +  .   \n",
       "4        chr1   HAVANA        exon  13221  14409  .      +  .   \n",
       "...       ...      ...         ...    ...    ... ..    ... ..   \n",
       "1878198  chrM  ENSEMBL  transcript  15888  15953  .      +  .   \n",
       "1878199  chrM  ENSEMBL        exon  15888  15953  .      +  .   \n",
       "1878200  chrM  ENSEMBL        gene  15956  16023  .      -  .   \n",
       "1878201  chrM  ENSEMBL  transcript  15956  16023  .      -  .   \n",
       "1878202  chrM  ENSEMBL        exon  15956  16023  .      -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        gene_id \"ENSG00000223972.5_1\"; gene_type \"tran...  \n",
       "1        gene_id \"ENSG00000223972.5_1\"; transcript_id \"...  \n",
       "2        gene_id \"ENSG00000223972.5_1\"; transcript_id \"...  \n",
       "3        gene_id \"ENSG00000223972.5_1\"; transcript_id \"...  \n",
       "4        gene_id \"ENSG00000223972.5_1\"; transcript_id \"...  \n",
       "...                                                    ...  \n",
       "1878198  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...  \n",
       "1878199  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...  \n",
       "1878200  gene_id \"ENSG00000210196.2\"; gene_type \"Mt_tRN...  \n",
       "1878201  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...  \n",
       "1878202  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...  \n",
       "\n",
       "[1878203 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "renamed = result.rename(columns={0: \"chrom\", 1: \"source\", 2: \"type\",3:\"start\",4:\"end\", 6:\"strand\"})\n",
    "renamed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>5</th>\n",
       "      <th>strand</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_1\"; gene_type \"tran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5_1\"; gene_type \"unpr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243485.5_1\"; gene_type \"lncR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237613.2_1\"; gene_type \"lncR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>52473</td>\n",
       "      <td>53312</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000268020.3_1\"; gene_type \"unpr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878184</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198695.2_1\"; gene_type \"prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878189</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210194.1\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878192</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198727.2_1\"; gene_type \"prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878197</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878200</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62446 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom   source  type  start    end  5 strand  7  \\\n",
       "0        chr1   HAVANA  gene  11869  14409  .      +  .   \n",
       "12       chr1   HAVANA  gene  14404  29570  .      -  .   \n",
       "25       chr1   HAVANA  gene  29554  31109  .      +  .   \n",
       "33       chr1   HAVANA  gene  34554  36081  .      -  .   \n",
       "38       chr1   HAVANA  gene  52473  53312  .      +  .   \n",
       "...       ...      ...   ...    ...    ... ..    ... ..   \n",
       "1878184  chrM  ENSEMBL  gene  14149  14673  .      -  .   \n",
       "1878189  chrM  ENSEMBL  gene  14674  14742  .      -  .   \n",
       "1878192  chrM  ENSEMBL  gene  14747  15887  .      +  .   \n",
       "1878197  chrM  ENSEMBL  gene  15888  15953  .      +  .   \n",
       "1878200  chrM  ENSEMBL  gene  15956  16023  .      -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        gene_id \"ENSG00000223972.5_1\"; gene_type \"tran...  \n",
       "12       gene_id \"ENSG00000227232.5_1\"; gene_type \"unpr...  \n",
       "25       gene_id \"ENSG00000243485.5_1\"; gene_type \"lncR...  \n",
       "33       gene_id \"ENSG00000237613.2_1\"; gene_type \"lncR...  \n",
       "38       gene_id \"ENSG00000268020.3_1\"; gene_type \"unpr...  \n",
       "...                                                    ...  \n",
       "1878184  gene_id \"ENSG00000198695.2_1\"; gene_type \"prot...  \n",
       "1878189  gene_id \"ENSG00000210194.1\"; gene_type \"Mt_tRN...  \n",
       "1878192  gene_id \"ENSG00000198727.2_1\"; gene_type \"prot...  \n",
       "1878197  gene_id \"ENSG00000210195.2\"; gene_type \"Mt_tRN...  \n",
       "1878200  gene_id \"ENSG00000210196.2\"; gene_type \"Mt_tRN...  \n",
       "\n",
       "[62446 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset = renamed.query(\"type == 'gene'\").copy()\n",
    "subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset.loc[:, \"gene_type\"] = subset.apply(lambda x: re.findall(r\"gene_type ([^\\;]*)\", x[8])[0].strip(\"\\\"\"), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_protein = subset.query(\"gene_type == 'protein_coding'\").copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>5</th>\n",
       "      <th>strand</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186092.7_1\"; gene_type \"prot...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>6166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>134901</td>\n",
       "      <td>139379</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237683.5\"; gene_type \"protei...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>367640</td>\n",
       "      <td>368634</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235249.1\"; gene_type \"protei...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>621059</td>\n",
       "      <td>622053</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662.1_1\"; gene_type \"prot...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>859303</td>\n",
       "      <td>879955</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.13_1\"; gene_type \"pro...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>20652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878157</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000212907.2_1\"; gene_type \"prot...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878164</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198886.2_1\"; gene_type \"prot...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878178</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198786.2_1\"; gene_type \"prot...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878184</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198695.2_1\"; gene_type \"prot...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1878192</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198727.2_1\"; gene_type \"prot...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1140</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20082 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom   source  type   start     end  5 strand  7  \\\n",
       "48       chr1   HAVANA  gene   65419   71585  .      +  .   \n",
       "77       chr1  ENSEMBL  gene  134901  139379  .      -  .   \n",
       "133      chr1   HAVANA  gene  367640  368634  .      +  .   \n",
       "197      chr1   HAVANA  gene  621059  622053  .      -  .   \n",
       "399      chr1   HAVANA  gene  859303  879955  .      +  .   \n",
       "...       ...      ...   ...     ...     ... ..    ... ..   \n",
       "1878157  chrM  ENSEMBL  gene   10470   10766  .      +  .   \n",
       "1878164  chrM  ENSEMBL  gene   10760   12137  .      +  .   \n",
       "1878178  chrM  ENSEMBL  gene   12337   14148  .      +  .   \n",
       "1878184  chrM  ENSEMBL  gene   14149   14673  .      -  .   \n",
       "1878192  chrM  ENSEMBL  gene   14747   15887  .      +  .   \n",
       "\n",
       "                                                         8       gene_type  \\\n",
       "48       gene_id \"ENSG00000186092.7_1\"; gene_type \"prot...  protein_coding   \n",
       "77       gene_id \"ENSG00000237683.5\"; gene_type \"protei...  protein_coding   \n",
       "133      gene_id \"ENSG00000235249.1\"; gene_type \"protei...  protein_coding   \n",
       "197      gene_id \"ENSG00000284662.1_1\"; gene_type \"prot...  protein_coding   \n",
       "399      gene_id \"ENSG00000187634.13_1\"; gene_type \"pro...  protein_coding   \n",
       "...                                                    ...             ...   \n",
       "1878157  gene_id \"ENSG00000212907.2_1\"; gene_type \"prot...  protein_coding   \n",
       "1878164  gene_id \"ENSG00000198886.2_1\"; gene_type \"prot...  protein_coding   \n",
       "1878178  gene_id \"ENSG00000198786.2_1\"; gene_type \"prot...  protein_coding   \n",
       "1878184  gene_id \"ENSG00000198695.2_1\"; gene_type \"prot...  protein_coding   \n",
       "1878192  gene_id \"ENSG00000198727.2_1\"; gene_type \"prot...  protein_coding   \n",
       "\n",
       "          size  \n",
       "48        6166  \n",
       "77        4478  \n",
       "133        994  \n",
       "197        994  \n",
       "399      20652  \n",
       "...        ...  \n",
       "1878157    296  \n",
       "1878164   1377  \n",
       "1878178   1811  \n",
       "1878184    524  \n",
       "1878192   1140  \n",
       "\n",
       "[20082 rows x 11 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_protein.loc[:, \"size\"] = subset_protein[\"end\"] - subset_protein[\"start\"]\n",
    "subset_protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes_forward = subset_protein.query(\"strand == '+'\")\n",
    "all_genes_reverse = subset_protein.query(\"strand == '-'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes_forward[[\"chrom\", \"start\", \"end\"]].to_csv(\"../../data/hg19_genes/all_genes_forward.bed\", sep=\"\\t\", header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes_reverse[[\"chrom\", \"start\", \"end\"]].to_csv(\"../../data/hg19_genes/all_genes_reverse.bed\", sep=\"\\t\", header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes_forward[[\"chrom\", \"start\", \"start\"]].to_csv(\"../../data/hg19_genes/protein_coding_genes_forward_tss.bed\", sep=\"\\t\", header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes_reverse[[\"chrom\", \"end\", \"end\"]].to_csv(\"../../data/hg19_genes/protein_coding_genes_reverse_tss.bed\", sep=\"\\t\", header=None, index=False)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "0603368e324c6de0b5c43c47fc82b6d2fed28d51bd7b2dc5642745b3fab8f71a"
  },
  "kernelspec": {
   "display_name": "Python 3.7.5 ('ml')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
